src/main/java/com/codejoust/main/service/ProblemService.java

```diff
@@ -153,29 +159,89 @@ public ProblemDto editProblem(String problemId, ProblemDto updatedProblem) {
             problem.addTestCase(testCase);
         }
 
-        repository.save(problem);
+        // If the same tag name is added multiple times, throw error.
+        List<ProblemTagDto> updatedProblemTags = updatedProblem.getProblemTags();
+        if (!problemTagNamesUnique(updatedProblemTags)) {
+            throw new ApiException(ProblemError.DUPLICATE_TAG_NAME);
+        }
+
+        // The old problem tags associated with the problem.
+        List<ProblemTag> oldProblemTags = new ArrayList<>(problem.getProblemTags());
+        for (ProblemTagDto problemTagDto : updatedProblemTags) {
+            // Add the tag if the name already exists, or create a new one.
+            ProblemTag existingProblemTag = problemTagRepository.findTagByName(problemTagDto.getName());
+            if (existingProblemTag != null) {
+                // Add the problem tag or remove from old list to be removed.
+                if (oldProblemTags.contains(existingProblemTag)) {
+                    oldProblemTags.remove(existingProblemTag);
+                } else {
+                    problem.addProblemTag(existingProblemTag);
+                }
+            } else if (validProblemTagName(problemTagDto.getName())) {
+                ProblemTag problemTag = new ProblemTag();
+                problemTag.setName(problemTagDto.getName());
+                problem.addProblemTag(problemTag);
+            } else {
+                throw new ApiException(ProblemError.BAD_PROBLEM_TAG);
+            }
+        }
+
+        /**
+         * Remove all the old problem tags.
+         * The tags own the removal side of the relationship, and are
+         * updated here whereas the problem is updated below.
+         */
+        for (ProblemTag oldProblemTag : oldProblemTags) {
+            oldProblemTag.removeProblem(problem);
+        }
+
+        problemRepository.save(problem);
+
+        // Remove the old problem tags from the problem to update the DTO.
+        for (ProblemTag oldProblemTag : oldProblemTags) {
+            problem.removeProblemTag(oldProblemTag);
+        }
```

Comment:
I wonder if this is because of the bidirectional mapping which causes this to be required in order to make it show correctly? One thought I had is if we simply removed the ManyToMany tag from the `ProblemTag.java`, we might be able to avoid having to do all the syncing (e.g. `problem.addTag`, `tag.addProblem`) and it might be easier to manage + more efficient, similar to the existing ManyToMany relationship between Room and Problem (only present on the Room side)

The only thing with this would be that the `getProblemsByTag` method would have to change its implementation from the current `problemTagRepository.findTagByTagId(tagId).getProblems()` to something like `problemRepository.findAllByTag(tag)` or `problemRepository.findAllByTag_Name(tagName)`

Reply:
I can try this solution, for sure - I'll see if this works and comment the results here. 👍

Reply:
Ok, so I tried to make this change, and everything else worked but when I removed the `ManyToMany` annotation on the `ProblemTag` class with the `List<Problem>`, I could not start the application or run the `ProblemTests` (the service ones worked just fine, but I got the following respective errors):

**Running the Application Error**
<img width="1280" alt="Running the Application Error" src="https://user-images.githubusercontent.com/7987279/119179878-530ecd00-ba24-11eb-9e66-713cb8da3f1d.png">

**ProblemTests Error**
<img width="732" alt="ProblemTests Error" src="https://user-images.githubusercontent.com/7987279/119179849-48ecce80-ba24-11eb-83ed-505d5da8f7c9.png">

If you know what the error is I'm happy to look at this again, but otherwise, the current implementation should work.

Reply:
I think it has something to do with [this](https://stackoverflow.com/questions/4011472/mappedby-reference-an-unknown-target-entity-property)? To test it I pushed a commit to a dummy branch [here](https://github.com/CodeJoustHQ/main/commit/1f551282f65880cefd4e219e5221e20a9dba9bf2), and I think it works (tests pass and basic Postman tests seem to work), though I didn't rigorously test it at all 

Lmk if any of that works (main reason I wanted to check this is to see if it could address the tag dissociating thing you mentioned + simplify/speed up the code) - otherwise tho I agree the current implementation is good as is

Reply:
Ah you're probably right - wow, thank you for making that whole separate commit to test it! I'll test this and work on integrating it now.

Reply:
This seems to work great! I will mention it was difficult to test it through Postman given that the problem routes now need tokens to validate the user, but I saved the support@codejoust.co token in Postman and hopefully we'll be able to use that more in the future.